network/src/noise_wrapper/handshake.rs

```diff
@@ -0,0 +1,390 @@
+const EXPIRATION_TIMESTAMP: u64 = 60;
+
+/// hashmap to store client timestamps if a connection succeeds
+pub type AntiReplayTimestamp = HashMap<x25519::PublicKey, u64>;
+
+// Noise Wrapper
+// -------------
+// Noise by default is not aware of the above or lower protocol layers,
+// We thus need to build this wrapper around Noise to both:
+//
+// - fragment messages that need to be encrypted by noise (due to its maximum 65535-byte messages)
+// - understand how long noise messages we send and receive are,
+//   in order to pass them to the noise implementaiton
+//
+
+/// TODO: why do we expose this to the outer protocol?
+const NOISE_PROTOCOL: &[u8] = b"/noise_ik_25519_aesgcm_sha256/1.0.0";
+
+/// The Noise configuration to be used to perform a protocol upgrade on an underlying socket.
+pub struct NoiseWrapper(noise::NoiseConfig);
+
+impl NoiseWrapper {
+    /// Create a new NoiseConfig with the provided keypair
+    pub fn new(key: x25519::PrivateKey) -> Self {
+        Self(noise::NoiseConfig::new(key))
+    }
+
+    /// Perform a protocol upgrade on an underlying connection. In addition perform the noise IX
+    /// handshake to establish a noise session and exchange static public keys. Upon success,
+    /// returns the static public key of the remote as well as a NoiseSession.
+    // TODO(mimoo, philp9): this code could be inlined in transport.rs once the monolithic network is done
+    pub async fn upgrade_connection<TSocket>(
+        &self,
+        socket: TSocket,
+        origin: ConnectionOrigin,
+        anti_replay_timestamps: Arc<RwLock<AntiReplayTimestamp>>,
+        remote_public_key: Option<x25519::PublicKey>,
+        trusted_peers: Option<&Arc<RwLock<HashMap<PeerId, NetworkPeerInfo>>>>,
+    ) -> io::Result<(x25519::PublicKey, NoiseSession<TSocket>)>
+    where
+        TSocket: AsyncRead + AsyncWrite + Unpin,
+    {
+        // Perform protocol negotiation
+        // TODO: should we really be doing this here?
+        let (socket, proto) = match origin {
+            ConnectionOrigin::Outbound => {
+                negotiate_outbound_interactive(socket, [NOISE_PROTOCOL]).await?
+            }
+            ConnectionOrigin::Inbound => negotiate_inbound(socket, [NOISE_PROTOCOL]).await?,
+        };
+
+        // check that the correct protocol was negotiated
+        if proto != NOISE_PROTOCOL {
+            if cfg!(test) {
+                panic!("noise: incorrect protocol negotiated");
+            } else {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    "noise: incorrect protocol negotiated",
+                ));
+            }
+        }
+
+        // perform the noise handshake
+        let socket = match origin {
+            ConnectionOrigin::Outbound => {
+                let remote_public_key = match remote_public_key {
+                    Some(key) => key,
+                    None if cfg!(test) => unreachable!(),
```

Comment:
is this so errors will panic in tests/fuzzing? probably use `panic` instead of `unreachable`, or consider removing this.

Reply:
actually only in tests, but yeah I'd want that in fuzzing too I guess. This should be unreachable though, why not use unreachable? That should give us a pretty good warning if we ever reach that panic